fendermint/vm/topdown/src/sync.rs

```diff
@@ -0,0 +1,206 @@
+// Copyright 2022-2023 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+//! A constant running process that fetch or listener to parent state
+
+use crate::error::Error;
+use crate::{BlockHash, BlockHeight, Config, InMemoryFinalityProvider, ParentFinalityProvider};
+use anyhow::{anyhow, Context};
+use async_stm::atomically_or_err;
+use fvm_shared::clock::ChainEpoch;
+use ipc_agent_sdk::apis::IpcAgentClient;
+use ipc_agent_sdk::jsonrpc::JsonRpcClientImpl;
+use ipc_agent_sdk::message::ipc::ValidatorSet;
+use ipc_sdk::cross::CrossMsg;
+use ipc_sdk::subnet_id::SubnetID;
+use std::sync::Arc;
+use std::time::Duration;
+
+/// Constantly syncing with parent through polling
+pub struct PollingParentSyncer {
+    config: Config,
+    parent_view_provider: Arc<InMemoryFinalityProvider>,
+    agent: Arc<IPCAgentProxy>,
+}
+
+impl PollingParentSyncer {
+    pub fn new(
+        config: Config,
+        parent_view_provider: Arc<InMemoryFinalityProvider>,
+        agent: Arc<IPCAgentProxy>,
+    ) -> Self {
+        Self {
+            config,
+            parent_view_provider,
+            agent,
+        }
+    }
+}
+
+impl PollingParentSyncer {
+    /// Start the parent finality listener in the background
+    pub fn start(self) -> anyhow::Result<()> {
+        let config = self.config;
+        let provider = self.parent_view_provider;
+        let agent = self.agent;
+
+        tokio::spawn(async move {
+            loop {
+                if let Err(e) = sync_with_parent(&config, &agent, &provider).await {
+                    tracing::info!("sync with parent encountered error: {e}");
+                }
+            }
+        });
+
+        Ok(())
+    }
+}
+
+async fn sync_with_parent<T: ParentFinalityProvider + Send + Sync + 'static>(
+    config: &Config,
+    agent_proxy: &Arc<IPCAgentProxy>,
+    provider: &Arc<T>,
+) -> anyhow::Result<()> {
+    let mut interval = tokio::time::interval(Duration::from_secs(config.polling_interval_secs));
+
+    loop {
+        interval.tick().await;
+
+        let (starting_height, latest_height) =
+            get_query_block_heights(agent_proxy, provider).await?;
+        if latest_height < config.chain_head_delay {
+            tracing::debug!("latest height not more than the chain head delay");
+            continue;
+        }
+
+        tracing::debug!("starting height: {starting_height}, latest_height: {latest_height}");
+
+        // we are going backwards in terms of block height, the latest block height is lower
+        // than our previously fetched head. It could be a chain reorg. We clear all the cache
+        // in `provider` and start from scratch
+        if starting_height > latest_height {
+            todo!()
+        }
+
+        let new_parent_views =
+            get_new_parent_views(agent_proxy, starting_height, latest_height).await?;
```

Comment:
Okay so we say we go form the latest height we know of + 1, and fetch until the latest height we see, then feed the blocks in this new range into our cache. Then, when we are asked what is final, we look into that cache and subtract 10 and call that block final.

The problem is that there _can_ be reorgs at the tip which this algorithm will fail to detect because of the +1 thing.

Reply:
Like I mentioned before I saw two options here:

#### Only fetch finalized blocks
1. Get the height of the parent chain
2. Subtract N to get the finalized height
3. Fetch blocks between our last fetch and the new finalized height. 

#### Deal with reorgs
1. Get the height of the parent chain
2. Fetch _again_ the block hash at the last height we fetched before
3. If it's the same, keep fetching until the latest height and append the new ones to the ache
4. If it's not the same, keep going back until we find a height where we still agree, and replace the ones in the cache before appending new ones.

Since we generally don't have to do anything with pending block, we might as well not fetch them, and have a dedicated `TVar` for the latest height, rather than get it as the height of the highest block we fetched.

Reply:
Nice one!

I have updated the ending height from `latest` chain head to `latest` chain head - N. But addressing the reorgs, reorg detection + error handling, I feel we might have to open another issue to address it.

I think we have two cases for reorgs, one is for the parent, one is for the child, i.e. fendermint.

For dealing with the cache in both scanarios, we can just clear cache and restart. The way we are garbage collection cache is when finality is committed, I reckon the overall data stored in cache should be small. Clearing the cache should be a `brutal` and simple way to handle this. I think the way you described can be used to detect the parent block height where the reorg happens, as the current check might be too simple. Also we might have to consider the case where the polling is just starting and the cache is empty, but parent reorg has already happened.

But I'm afraid the more complex issue is revert the execution of top down messages and revert the committed validator set.

That's why I feel we should start another issue to address the reorg while this PR just ensures the happy path works first.

Reply:
> But I'm afraid the more complex issue is revert the execution of top down messages and revert the committed validator set.

Well, we should never have to revert the execution of top down messages and committed validator sets, because we should not execute them until they are final. I know you added this default delay of 10 blocks, but to be correct, before something like the Granite work is added to Filecoin, we must actually use 900 blocks if the parent is Lotus. 

> I think we have two cases for reorgs, one is for the parent, one is for the child, i.e. fendermint.

If you mean what if Fendermint reorgs, you don't have to worry, it never does.